pkgs/by-name/wi/wivrn/package.nix

```diff
@@ -0,0 +1,122 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, fetchgit
+, avahi
+, boost
+, cmake
+, cudaPackages
+, eigen
+, ffmpeg
+, freetype
+, git
+, glm
+, glslang
+, harfbuzz
+, libdrm
+, libva
+, libpulseaudio
+, monado
+, nlohmann_json
+, onnxruntime
+, openxr-loader
+, pkg-config
+, python3
+, shaderc
+, spdlog
+, systemd
+, udev
+, vulkan-headers
+, vulkan-loader
+, vulkan-tools
+, x264
+, xorg
+}:
+let
+  vendorMonado = monado.overrideAttrs rec {
+    # Version stated in CMakeList for WiVRn 0.11
+    version = "57e937383967c7e7b38b5de71297c8f537a2489d";
+
+    src = fetchgit {
+      url = "https://gitlab.freedesktop.org/monado/monado.git";
+      rev = "${version}";
+      hash = "sha256-O/Td2WccTr4Fa8U64/lVfnidSIH5t3gWuFXCsEVf7bk=";
+    };
+
+    postInstall = ''
+      mv src/xrt/compositor/libcomp_main.a $out/lib/libcomp_main.a
+    '';
+  };
+in
+stdenv.mkDerivation rec {
+  pname = "wivrn";
+  version = "0.11";
+
+  src = fetchFromGitHub {
+    owner = "meumeu";
+    repo = "wivrn";
+    rev = "v${version}";
+    hash = "sha256-E48JYhkBVZfb7S/FW0F8RbMtx4GIJwfXfs4KAF3gn8A=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    cudaPackages.cuda_nvcc
+    pkg-config
+    python3
+  ];
+
+  buildInputs = [
+    avahi
+    boost
+    cudaPackages.cuda_cudart
+    eigen
+    ffmpeg
+    freetype
+    git
+    glm
+    glslang
+    harfbuzz
+    libdrm
+    libva
+    libpulseaudio
+    nlohmann_json
+    onnxruntime
+    openxr-loader
+    shaderc
+    spdlog
+    systemd
+    udev
+    vendorMonado
+    vulkan-headers
+    vulkan-loader
+    vulkan-tools
+    x264
+    xorg.libX11
+    xorg.libXrandr
+  ];
+
+  cmakeFlags = [
+    "-DWIVRN_BUILD_CLIENT=OFF"
+    "-DCMAKE_BUILD_TYPE=RelWithDebInfo"
+    "-DFETCHCONTENT_FULLY_DISCONNECTED=ON"
+    "-DFETCHCONTENT_SOURCE_DIR_MONADO=${vendorMonado.src}"
+    "-DWIVRN_USE_VAAPI=ON"
+    "-DWIVRN_USE_X264=ON"
+    "-DWIVRN_USE_NVENC=OFF"
+  ];
+
+  postFixup = ''
+    substituteInPlace $out/share/openxr/1/openxr_wivrn.json --replace "../../../" "../../../../.."
```

Comment:
I tried github codesearch on the repo but that didn't help. What's the structure of this file, what does `../../..` refer to? Is there any possibility that upstream might incorporate the same change (`../`*3 -> `../`*4)? If that ever happens, this substitution will result in `../`*5.

I wonder if it would make more sense to format an actual `.patch`, so that its application would fail when the context changes. Is this even important?

Reply:
The file it creates should look like:
```json
{
  "file_format_version": "1.0.0",
  "runtime": {
    "library_path": "../../../../../nix/store/cm8dkr5nw1ci3146q6sbvhhp2mx8li05-wivrn-0.11/lib/libopenxr_wivrn.so"
  }
}
```
where the nix store link is relative to the libopenxr file.

However, without fixing it up, it will look like `../../../..//nix/store/...` which is invalid.

Reply:
LMAO. So we're doing something like `s|(../)*${builtins.storeDir}|${builtins.storeDir}|g`

Reply:
> I think we should patch this in the source code instead

Thanks for locating this! Yes, absolutely. We probably should reach out with the upstream because this means their cmake "package" isn't [relocatable](https://cmake.org/cmake/help/latest/guide/importing-exporting/index.html#creating-relocatable-packages)

Reply:
Currently I just moved the patch from postFixup to patchPhase. I tried to just remove that line that Scrumplex mentioned but it still seemed to include it in the output. I can ask upstream if this is something that can be changed, although it might serve a purpose somewhere else.

Reply:
I just tested it by removing the relative paths entirely so its just a store path and it actually appears to work fine. Not sure if that is intentional however

Reply:
Tracking in https://github.com/Meumeu/WiVRn/issues/47. Could you please attach a brief human-redable comment and a link just above the patch?

Reply:
The OpenXR loader will process the json file and is described at https://github.com/KhronosGroup/OpenXR-SDK-Source/blob/main/specification/loader/runtime.adoc

The loader will search in xdg configuration directories, and only one active runtime can be configured at a given time. WiVRn package should own the json file, and I normally install it in `/usr/share/openxr/1/openxr_wivrn.json`, then a "configuration tool", which does not exist yet on Gentoo configures the symlink in `/etc/openxr/1/active_runtime.json` to point to the WiVRn one. Because this tool does not exist, the package install just creates it, this should be removed if we want to allow a user to install multiple runtimes and provide some mechanism to select the active one.

Since https://github.com/Meumeu/WiVRn/commit/9efe3852f4c6fc35e39bef6a40ef375145f1261b, the wivrn-server process will try to set itself as user default when a connection is established by creating the symlink to `WIVRN_INSTALL_PREFIX/share/openxr/1/openxr_wivrn.json`

Reply:
I've attempted to place the `active_runtime.json` under `/etc/openxr/1/active_runtime.json` and `/etc/xdg/openxr/1/active_runtime.json` but it was never found. It was, however, found at `~/.config/openxr/1/active_runtime.json`.

I believe this is because of the steam Pressure Vessel filesystem but I'm not quite sure how to resolve it.

Reply:
It turns out that OpenXR can use an absolute path to the runtime so the relative paths are not needed and can be removed entirely. I'll leave this review unresolved however as there is more going on.

I think all that's left to figure out is where to generate the `active_runtime.json` so it can be found by Steam